tidy3d/plugins/near2far/near2far.py

```diff
@@ -1,115 +1,373 @@
+
+        currents.data_dict['M'+tangent_fields[0]].values *= signs[1]
+        currents.data_dict['M'+tangent_fields[1]].values *= signs[0]
+
+        # colocate surface currents on a regular grid of points on the monitor based on wavelength
+        colocation_points = [None] * 3
+        colocation_points[self.axis] = self.mon.center[self.axis]
+
+        for idx in idx_uv:
+
+            # pick sample points on the monitor and handle the possibility of an "infinite" monitor
+            start = np.maximum(
+                self.mon.center[idx] - self.mon.size[idx] / 2.0,
+                self.mon.center[idx] - self.sim_data.simulation.size[idx] / 2.0
+                )
+            stop = np.minimum(
+                self.mon.center[idx] + self.mon.size[idx] / 2.0,
+                self.mon.center[idx] + self.sim_data.simulation.size[idx] / 2.0
+                )
+            size = stop - start
+
+            num_pts = int(np.ceil(pts_per_wavelength * size / wavelength))
+            points = np.linspace(start, stop, num_pts)
+            colocation_points[idx] = points
+
+        currents = currents.colocate(*colocation_points)
         try:
-            Ex = field_data.Ex.sel(f=frequency)
-            Ey = field_data.Ey.sel(f=frequency)
-            # self.Ez = field_data['Ez'].sel(f=frequency)
-            Hx = field_data.Hx.sel(f=frequency)
-            Hy = field_data.Hy.sel(f=frequency)
+            self.currents = currents.sel(f=frequency)
         except Exception as e:
             raise SetupError(
-                f"Frequency {frequency} not found in all fields " f"from monitor '{mon_name}'."
+                f"Frequency {frequency} not found in fields for monitor '{self.mon.name}'."
             ) from e
 
-        self.k0 = 2 * np.pi * frequency / C_0
+class Near2Far(Tidy3dBaseModel):
+    """Near field to far field transformation tool.
+
+    Parameters
+    ----------
+    sim_data : :class:`.SimulationData`
+        Container for simulation data containing the near field monitors.
+    surfaces : List[:class:`Near2FarSurface`]
+        List of each :class:`.Near2FarSurface` to use as source of near field.
+    frequency : float
+        Frequency to select from each :class:`.FieldMonitor` to use for projection.
+        Must be a frequency stored in each :class:`FieldMonitor`.
+    pts_per_wavelength : int
+        Number of points per wavelength with which to discretize the
+        surface monitors for the projection. Default: 10.
+    medium : :class:`.Medium`
+        Background medium in which to radiate near fields to far fields.
+        Default: same as the :class:`.Simulation` background medium.
+    """
+
+    sim_data: SimulationData
+    surfaces: List["Near2FarSurface"]
+    frequency: float
+    pts_per_wavelength: int = 10
+    medium: Medium = None
+
+    origin: List[float] = [0, 0, 0]
+    phasor_sign: float = 1  # 1 => exp(jkr), -1 => exp(-jkr)
```

Comment:
We probably dont need to expose this, just hardcode it

Reply:
yea you can put an underscore in front of the field name and it will be hidden, not written to file, not documented.  but still set if you want to (I think)